lib/ld-eventsource/client.rb

```diff
@@ -90,15 +90,20 @@ def initialize(uri,
           reconnect_reset_interval: DEFAULT_RECONNECT_RESET_INTERVAL,
           last_event_id: nil,
           proxy: nil,
-          logger: nil)
+          logger: nil,
+          socket_factory: nil)
```

Comment:
We need a doc comment for the new parameter. Yard doesn't really have a way to describe an interface-like type, since Ruby doesn't either, but it can at least describe the use of duck-taping to require a single method— and in this case that's what this is, as the only requirement for the factory object is that it implement `open(host, port)`. I suggest:
```
# @param socket_factory [#open] (nil)  an optional factory object for creating sockets, if you want to use something other than the default `TCPSocket`; it must implement `open(host, port)` to return a `Socket`
```

Looking at the httprb source, I see that there's a separate option for `ssl_socket_class`. This doesn't have the same semantics as `socket_class`— it's used to add an encryption layer after the original socket has already been opened. If the stream URL might still have an `https:` scheme even we're using Unix sockets, and if the default implementation of SSL over a TCP socket won't work for Unix sockets, then we need to account for this.

Reply:
(Unfortunately `[#open(host, port)]` is not something Yard understands.)

Reply:
The Ruby OpenSSL library just takes a Ruby IO Object and I think it should work though I can test it if you want. OpenSSL normally does work with Unix Sockets. I expect utilizing Unix sockets *and* HTTPS is going to be unlikely though.

I will update the doc string as described.

Reply:
> I expect utilizing Unix sockets and HTTPS is going to be unlikely though.

OK, maybe I'm remembering wrong but I thought you said at some point that people might still be using the original service URLs, like `https://app.launchdarkly.com`, if the thing that they are talking to through the Unix sockets behaves more or less like a reverse proxy. If so, it's immaterial whether we really want it to be doing TLS over the Unix socket; httprb will look at that URL and decide that it needs to use `ssl_socket_class`.

Reply:
Well yeah but I *expect* it to use TLS if you use `https://myrelay.internal`.

Reply:
But, assuming there is such a reverse-proxy-type service, is that how it expects things to work? Is it even relevant when you're talking about interprocess communications that aren't going over a wire at that point?

Maybe this is just my patchy knowledge of TLS showing, but it seems to me that in order to tunnel TLS over a Unix socket in that way, the proxy service would have to have its own certificate— and it would have to know which hostname is being requested without being able to see the unencrypted request. Secure web proxy servers deal with this by having a CONNECT command that gets sent _before_ encryption starts, but that requires the client to know that it's talking to a proxy server, which it wouldn't in this case.

So, if we're expecting a proxy thing on the other end of the Unix socket to look at the hostname and connect securely to either `https://stream.launchdarkly.com` or `https://events.launchdarkly.com`, etc., I would think it has to be up to that proxy to implement TLS on the communication that's actually going over the wire.

Reply:
In other words, what I was imagining in this scenario was:

1. SDK wants to make a request to `https://stream.launchdarkly.com`
2. Httprb client sends a plain HTTP request over the Unix socket, with that hostname in the Host header
3. Proxy service checks the Host header and issues the HTTPS request to that host

But if you allow the httprb client to apply TLS in the way it will do by default, I think this would happen:

1. SDK wants to make a request to `https://stream.launchdarkly.com`
2. Httprb client opens the Unix socket and tries to start a TLS handshake
3. Proxy service doesn't know what certificate to use for this, because it can't tell what host the client wants to connect to. Or: proxy service uses its own certificate, but handshake fails because the cert is not for the desired host.

Reply:
Again, I'm just imagining a kind of service that someone might be using, but I have no idea whether that's actually a thing or whether the customer in this case is using such a thing.

Reply:
There is a part of TLS called SNI that let's proxies / servers know the domain to proxy to, or certificate it should use. The socket factory could also technically do arbitrary setup work by sending a `CONNECT` or something. But either way what you are describing is not really the target use case.

Reply:
@hroederld So am I remembering wrong that, in the target use case, the LD service URLs would be the usual ones? If there was such a conversation, I don't know where it happened— there's no mention of the URLs on the "Ruby SDK Unix socket support" Confluence page— but I had a strong feeling that we had talked about it.

Reply:
Hmm, well, I do see this: https://launchdarkly.atlassian.net/wiki/spaces/~104190358/pages/826573367/Square+-+Abstract+Unix+Sockets+FR?focusedCommentId=847773957#comment-847773957 —which I think implies that I was wrong, and that the URL would be something like `http://production.ld-relay.gns.square` (though you didn't mention a URL scheme).

Reply:
(I did remember wrong - the customer is targeting a Relay instance rather than the real LD service URLs.)